### PR TITLE
select graph needs based on method args for more flexibility

### PR DESCRIPTION
**Pull request by tmbo**
- error if arguments are not met
- rather than fixed needs, needs are based on the arguments of the graph function

### CODE REVIEW

File: rasa/engine/graph.py [hunk 1]
```diff
@@ -436,10 +440,25 @@ def __call__(
         Returns:
             The node name and its output.
         """
-        received_inputs: Dict[Text, Any] = dict(inputs_from_previous_nodes)
+        # filter out arguments that dask couldn't lookup
+        received_inputs: Dict[Text, Any] = {}
+        for i in inputs_from_previous_nodes:
+            if type(i) is tuple:
```
Comment (twerkmeister) [inline rasa/engine/graph.py:446]: isinstance

File: rasa/engine/graph.py [hunk 2]
```diff
@@ -436,10 +440,25 @@ def __call__(
...
+                    f"Received unresolved argument: '{i}'. "
+                    f"Another component should have provided this as an "
+                    f"output."
+                )
 
         kwargs = {}
         for input_name, input_node in self._inputs.items():
+            if input_node not in received_inputs:
+                raise GraphRunError(
+                    f"Missing input from parent node. Expected input '{input_node}'."
```
Comment (twerkmeister) [inline rasa/engine/graph.py:460]: expected input input_name from input_name in current_node_name?

**Review by twerkmeister — CHANGES_REQUESTED** [review]
some initial proposals

File: rasa/engine/graph.py [hunk 3]
```diff
@@ -436,10 +440,27 @@ def __call__(
...
+        for i in inputs_from_previous_nodes:
+            if isinstance(i, tuple):
+                arg_name, arg_value = i
+                received_inputs[arg_name] = arg_value
+            else:
+                logger.warning(
+                    f"Received unresolved argument: '{i}'. "
+                    f"Another component should have provided this as an "
+                    f"output."
+                )
```
Comment (twerkmeister) [inline rasa/engine/graph.py:454]: Something seems off about this section 🤔. At least the naming should be improved.

1. `inputs_form_previous_nodes` is a mapping of node name to output of a node according to the doc string of `__call__`.
2. in line 447, an element of this mapping is deconstructed into `arg_name` and `arg_value`. Judging from the previous observation this should be probably `node_name` and `node_output`
3. In line 451 a warning is issued using the `Text` type inputs from previous node as the name of an argument when the other elements and the doc string hint at the items in this list to be `node_name` and `node_output` pairs.

File: rasa/engine/graph.py [hunk 4]
```diff
@@ -436,10 +440,27 @@ def __call__(
...
 
         kwargs = {}
         for input_name, input_node in self._inputs.items():
+            if input_node not in received_inputs:
+                raise GraphRunError(
+                    f"Missing input to run node '{self._node_name}'. "
+                    f"Expected input '{input_node}' for parameter "
+                    f"'{input_name}'."
+                )
```
Comment (twerkmeister) [inline rasa/engine/graph.py:463]: Maybe also some potential for improving the naming and the error message here.
 I think `input_node` should probably be `input_provider_node_name` and then in the second part of the error message you could say "Expected '`input_provider_node_name`' to provide parameter '`input_name`'."

**Review by twerkmeister — CHANGES_REQUESTED** [review]
The changes seem innocent enough, and I can only spot some naming issues right now. With regards to the failing test, however, there seems to be another issue. The test introduces a tokenizer into the policy graph and expects the graph validation to fail. It seems this failure is not happening and instead the policy is being trained which then results in the 60s timeout happening in some tensorflow code. Wasn't able to really get why the changed code wouldn't provide a failure when it did before. Maybe you have an idea.

File: rasa/engine/graph.py (hunk 3 above)
Comment (tmbo) [reply rasa/engine/graph.py:454]: sounds good, renamed the args. yes the description isn't entirely correct. as you can see I changed the type annotation as for node names that couldn't be resolved there is no tuple but just the node name.

**Review by twerkmeister — APPROVED** [review]

**Comment by sutgeorge** [comment]
I have this error popping up, even if I implemented my custom intent classifier as expected. Can you please give some details on what this exception is supposed to mean?

```
packages/rasa/engine/validation.py", line 173, in _walk_and_check_for_cycles
    raise GraphSchemaValidationException(
rasa.engine.exceptions.GraphSchemaValidationException: Node 'train_data_provider' is not part of the graph. Node was expected to be present in the graph as it is used by another component.
```

**Comment by sutgeorge** [comment]
I found this information specified in the documentation, hopefully this is my actual issue. I used an implementation that was probably suitable for Rasa 2, but not for Rasa 3: https://rasa.com/docs/rasa/migration-guide#changes-to-custom-nlu-components

**Comment by sutgeorge** [comment]
Alright, I'm back. Sorry for the ping @koaning , but is it possible to include these examples of a custom intent classifier into the documentation? (at least the links to the repository would be incredibly useful, I wasted quite a lot of time on wrestling with this exception).

- https://github.com/RasaHQ/rasa-nlu-examples/blob/main/rasa_nlu_examples/classifiers/sparse_naive_bayes_intent_classifier.py
- https://github.com/RasaHQ/rasa-nlu-examples/blob/fe12dbc814d992382c1ca1d926b340139200928f/tests/configs/sparse-naive-bayes-intent-classifier-config.yml

**Pull request merged**